src/amo/components/InstallWarning/index.js

```diff
@@ -0,0 +1,184 @@
+    WithExperimentInjectedProps,
+    'experimentEnabled',
+  >,
+  installStatus: $PropertyType<InstalledAddon, 'status'>,
+|};
+
+export const couldShowWarning = ({
+  addonIsExtension,
+  addonIsRecommended,
+  experimentEnabled,
+  installStatus,
+}: CouldShowWarningParams) => {
+  return (
+    addonIsExtension &&
+    !addonIsRecommended &&
+    experimentEnabled &&
+    [UNINSTALLED, UNKNOWN].includes(installStatus)
+  );
+};
+
+const addonIsExtension = (addon: AddonType) => {
+  return addon.type === ADDON_TYPE_EXTENSION;
+};
+
+type InternalProps = {|
+  ...Props,
+  ...WithExperimentInjectedProps,
+  _couldShowWarning: typeof couldShowWarning,
+  _log: typeof log,
+  _tracking: typeof tracking,
+  i18n: I18nType,
+  installStatus: $PropertyType<InstalledAddon, 'status'>,
+|};
+
+export const EXPERIMENT_CATEGORY_CLICK =
+  'AMO Install Button Warning Experiment - Click';
+export const EXPERIMENT_CATEGORY_SHOW =
+  'AMO Install Button Warning Experiment - Display';
+export const EXPERIMENT_ID = 'installButtonWarning';
+export const INSTALL_WARNING_EXPERIMENT_DIMENSION = 'dimension6';
+export const VARIANT_INCLUDE_WARNING = 'includeWarning';
+export const VARIANT_EXCLUDE_WARNING = 'excludeWarning';
+
+export class InstallWarningBase extends React.Component<InternalProps> {
+  static defaultProps = {
+    _couldShowWarning: couldShowWarning,
+    _log: log,
+    _tracking: tracking,
+  };
+
+  maybeSendDisplayTrackingEvent = () => {
+    const {
+      _couldShowWarning,
+      _tracking,
+      addon,
+      experimentEnabled,
+      installStatus,
+      variant,
+    } = this.props;
+
+    if (
+      _couldShowWarning({
+        addonIsExtension: addonIsExtension(addon),
+        addonIsRecommended: addon.is_recommended,
+        experimentEnabled,
+        installStatus,
+      }) &&
+      variant
+    ) {
+      _tracking.sendEvent({
+        action: variant,
+        category: EXPERIMENT_CATEGORY_SHOW,
+        label: addon.name,
+      });
+    }
+  };
+
+  componentDidMount() {
+    const { _log, _tracking, variant } = this.props;
+
+    if (!variant) {
+      _log.debug(`No variant set for experiment "${EXPERIMENT_ID}"`);
+      return;
+    }
+
+    _tracking.setDimension({
+      dimension: INSTALL_WARNING_EXPERIMENT_DIMENSION,
+      value: variant,
+    });
+
+    this.maybeSendDisplayTrackingEvent();
+  }
+
+  componentDidUpdate({ installStatus: oldInstallStatus }: InternalProps) {
+    const { installStatus: newInstallStatus } = this.props;
+
+    if (newInstallStatus !== oldInstallStatus) {
+      this.maybeSendDisplayTrackingEvent();
+    }
+  }
+
+  render() {
+    const {
+      _couldShowWarning,
+      addon,
+      experimentEnabled,
+      i18n,
+      installStatus,
+      variant,
+    } = this.props;
+
+    if (
+      _couldShowWarning({
+        addonIsExtension: addonIsExtension(addon),
+        addonIsRecommended: addon.is_recommended,
+        experimentEnabled,
+        installStatus,
+      }) &&
```

Comment:
Seeing that `_couldShowWarning` is called twice in the same class makes me think that we should probably have a method that wraps it. Or perhaps, `_couldShowWarning` should not be a helper but a method in `InstallWarning`.

Reply:
I don't totally follow what your concern is. I actually started by having it as a class method, but I wanted to be able to mock it out, and making it a helper seemed like an easier way of doing that. I'm not sure how having it as a class method would be any better at this point. I'm also not sure what you mean by having a method that wraps it - wouldn't we still have to pass all of the values into the wrapper, and therefore how would that be an improvement?

Reply:
I don't totally follow what your concern is. I actually started by having it as a class method, but I wanted to be able to mock it out, and making it a helper seemed like an easier way of doing that. I'm not sure how having it as a class method would be any better at this point. I'm also not sure what you mean by having a method that wraps it - wouldn't we still have to pass all of the values into the wrapper, and therefore how would that be an improvement?

Reply:
Most arguments given to `_couldShowWarning` are props so having a method helps with that because you would not have to repeat the call to `_couldShowWarning` with a long list of arguments, you would only call `this.couldShowWarning()`.

Reply:
Did a comment of mine get discarded? I wrote something in response to this but I don't see it in the thread now. The reason I made it a helper and not a class method was so I could mock it out.

Reply:
Yeah, I've seen it, then it was another one, then I replied, then it was the original one, and no... nothing.

Reply:
Heh, I don't really understand that last comment, the one that ends with "and no... nothing", but I think GitHub is being weird.

I agree it would be better to have this as a class method. I think I'll do that and get rid of the tests I have for `couldShowWarning` and the tests that use a mock for it, and just make them tests against the component and allow it to call `couldShowWarning` internally.

Reply:
The problem with not having a set of tests for `couldShowWarning` is that it's used in two different places, and it has a lot of different data conditions that it needs to be tested with. So if I don't test it in isolation, then I'm going to end up with double the number of tests (one set for the tracking event and one set for displaying the warning), so I guess it is better to keep the same model where I test it in isolation and then mock it. Testing it in isolation requires more setup now that it uses props rather than arguments, but I think that's still for the best.